hudi-common/src/main/java/org/apache/hudi/common/table/log/TypeCaster.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.common.table.log;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Type;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TypeCaster {
```

Comment:
How is this different from `SparkFileFormatInternalRowReaderContext#castValue` (which is also engine-independent)? Should we only have one unified type cast logic without code duplication?

Reply:
Yeah, we can move it to some utils if needed.

Reply:
Just checked, UTF8String is from spark, which is engine specific, so we cannot directly use it.

Reply:
We can directly use and and then add utf8 ontop of the string in SparkFileFormatInternalRowReaderContext. We will call tostring anyways, so overhead should be negligible

Reply:
Here is what I mean:
```
  override def castValue(value: Comparable[_], newType: Schema.Type): Comparable[_] = {
    value match {
      case v: UTF8String => newType match {
        case Type.STRING => v
        case x => throw new UnsupportedOperationException(s"Cast from String to $x is not supported")
      }
      case _ => newType match {
        case Type.STRING => UTF8String.fromString(TypeCaster.castValue(value, newType))
        case _ => TypeCaster.castValue(value, newType)
      }
    }
  }
```

Reply:
@linliu-code let's follow @jonvex 's suggestion and leverage `TypeCaster` in `SparkFileFormatInternalRowReaderContext` to avoid code duplication, so that only String is handled independently.

Reply:
Done.